BLD-591 - ps-admin.sh script for RocksDB/TokuDB and plugins easy install

tplavcic opened the pull request:
**Ticket:** BLD-591

What was done:
- changed some specific options from ps_tokudb_admin like
`--enable` -> `--enable-tokudb,` `--enable-backup` -> `--enable-tokubackup` and similar, and also changed variable names to not be tokudb specific
- added setup of RocksDB
- added setup of plugins: Audit Log, PAM, Query Response Time, MYSQLX
- added `--config-file` option
- added into CMakeLists.txt that the script is installed only if built with -DWITH_ROCKSDB since currently we don't want this script to be shipped

**TEST:**
**RocksDB:**
```
plavi@zoidberg  ~/sandboxes/binaries/5.7.17/bin  sudo ./percona_server_admin --enable-rocksdb --user=msandbox --password=msandbox --socket=/tmp/mysql_sandbox5717.sock                                                            
Checking RocksDB engine plugin status...
INFO: RocksDB engine plugin is not installed.

Installing RocksDB engine...
INFO: Successfully installed RocksDB engine plugin.

mysql [localhost] {msandbox} ((none)) > show engines;
+--------------------+-----------+----------------------------------------------------------------------------+----------------+--------+--------------+
| Engine             | Support   | Comment                                                                    | Transactions   | XA     | Savepoints   |
|--------------------+-----------+----------------------------------------------------------------------------+----------------+--------+--------------|
| ROCKSDB            | YES       | RocksDB storage engine                                                     | YES            | YES    | YES          |
+--------------------+-----------+----------------------------------------------------------------------------+----------------+--------+--------------+

+-----------------------------+----------+--------------------+---------------+-----------+
| Name                        | Status   | Type               | Library       | License   |
|-----------------------------+----------+--------------------+---------------+-----------|
| ROCKSDB                     | ACTIVE   | STORAGE ENGINE     | ha_rocksdb.so | GPL       |
| ROCKSDB_CFSTATS             | ACTIVE   | INFORMATION SCHEMA | ha_rocksdb.so | GPL       |
| ROCKSDB_DBSTATS             | ACTIVE   | INFORMATION SCHEMA | ha_rocksdb.so | GPL       |
| ROCKSDB_PERF_CONTEXT        | ACTIVE   | INFORMATION SCHEMA | ha_rocksdb.so | GPL       |
| ROCKSDB_PERF_CONTEXT_GLOBAL | ACTIVE   | INFORMATION SCHEMA | ha_rocksdb.so | GPL       |
| ROCKSDB_CF_OPTIONS          | ACTIVE   | INFORMATION SCHEMA | ha_rocksdb.so | GPL       |
| ROCKSDB_GLOBAL_INFO         | ACTIVE   | INFORMATION SCHEMA | ha_rocksdb.so | GPL       |
| ROCKSDB_COMPACTION_STATS    | ACTIVE   | INFORMATION SCHEMA | ha_rocksdb.so | GPL       |
| ROCKSDB_DDL                 | ACTIVE   | INFORMATION SCHEMA | ha_rocksdb.so | GPL       |
| ROCKSDB_INDEX_FILE_MAP      | ACTIVE   | INFORMATION SCHEMA | ha_rocksdb.so | GPL       |
| ROCKSDB_LOCKS               | ACTIVE   | INFORMATION SCHEMA | ha_rocksdb.so | GPL       |
| ROCKSDB_TRX                 | ACTIVE   | INFORMATION SCHEMA | ha_rocksdb.so | GPL       |
+-----------------------------+----------+--------------------+---------------+-----------+

plavi@zoidberg  ~/sandboxes/binaries/5.7.17/bin  sudo ./percona_server_admin --disable-rocksdb --user=msandbox --password=msandbox --socket=/tmp/mysql_sandbox5717.sock                                                             
Checking RocksDB engine plugin status...
INFO: RocksDB engine plugin is installed.

Uninstalling RocksDB engine plugin...
INFO: Successfully uninstalled RocksDB engine plugin.
```

**Query Response Time:**
```
ubuntu@t-ubuntu1604-64:~$ sudo percona_server_admin --user=root --password=Tomi1234! --enable-qrt
Checking Query Response Time plugin status...
INFO: Query Response Time plugin is not installed.

Installing Query Response Time plugin...
INFO: Successfully installed Query Response Time plugin.

ubuntu@t-ubuntu1604-64:~$ mysql --user=root --password=Tomi1234! -e "select * from information_schema.plugins where plugin_name like '%response%';"
mysql: [Warning] Using a password on the command line interface can be insecure.
+---------------------------+----------------+---------------+--------------------+---------------------+------------------------+------------------------+------------------------------+------------------------------------------------------------+----------------+-------------+
| PLUGIN_NAME               | PLUGIN_VERSION | PLUGIN_STATUS | PLUGIN_TYPE        | PLUGIN_TYPE_VERSION | PLUGIN_LIBRARY         | PLUGIN_LIBRARY_VERSION | PLUGIN_AUTHOR                | PLUGIN_DESCRIPTION                                         | PLUGIN_LICENSE | LOAD_OPTION |
+---------------------------+----------------+---------------+--------------------+---------------------+------------------------+------------------------+------------------------------+------------------------------------------------------------+----------------+-------------+
| QUERY_RESPONSE_TIME_AUDIT | 1.0            | ACTIVE        | AUDIT              | 4.1                 | query_response_time.so | 1.7                    | Percona and Sergey Vojtovich | Query Response Time Distribution Audit Plugin              | GPL            | ON          |
| QUERY_RESPONSE_TIME       | 1.0            | ACTIVE        | INFORMATION SCHEMA | 50717.0             | query_response_time.so | 1.7                    | Percona and Sergey Vojtovich | Query Response Time Distribution INFORMATION_SCHEMA Plugin | GPL            | ON          |
| QUERY_RESPONSE_TIME_READ  | 1.0            | ACTIVE        | INFORMATION SCHEMA | 50717.0             | query_response_time.so | 1.7                    | Percona and Sergey Vojtovich | Query Response Time Distribution INFORMATION_SCHEMA Plugin | GPL            | ON          |
| QUERY_RESPONSE_TIME_WRITE | 1.0            | ACTIVE        | INFORMATION SCHEMA | 50717.0             | query_response_time.so | 1.7                    | Percona and Sergey Vojtovich | Query Response Time Distribution INFORMATION_SCHEMA Plugin | GPL            | ON          |
+---------------------------+----------------+---------------+--------------------+---------------------+------------------------+------------------------+------------------------------+------------------------------------------------------------+----------------+-------------+

ubuntu@t-ubuntu1604-64:~$ sudo percona_server_admin --user=root --password=Tomi1234! --disable-qrt
Checking Query Response Time plugin status...
INFO: Query Response Time plugin is installed.

Uninstalling Query Response Time plugin...
INFO: Successfully uninstalled Query Response Time plugin.
```

**Audit Log plugin:**
```
ubuntu@t-ubuntu1604-64:~$ sudo percona_server_admin --user=root --password=Tomi1234! --enable-audit
Checking Audit Log plugin status...
INFO: Audit Log plugin is not installed.

Installing Audit Log plugin...
INFO: Successfully installed Audit Log plugin.

ubuntu@t-ubuntu1604-64:~$ mysql --user=root --password=Tomi1234! -e "select * from information_schema.plugins where plugin_name like '%audit%';"
+-------------+----------------+---------------+-------------+---------------------+----------------+------------------------+------------------------------------+--------------------+----------------+-------------+
| PLUGIN_NAME | PLUGIN_VERSION | PLUGIN_STATUS | PLUGIN_TYPE | PLUGIN_TYPE_VERSION | PLUGIN_LIBRARY | PLUGIN_LIBRARY_VERSION | PLUGIN_AUTHOR                      | PLUGIN_DESCRIPTION | PLUGIN_LICENSE | LOAD_OPTION |
+-------------+----------------+---------------+-------------+---------------------+----------------+------------------------+------------------------------------+--------------------+----------------+-------------+
| audit_log   | 0.2            | ACTIVE        | AUDIT       | 4.1                 | audit_log.so   | 1.7                    | Percona LLC and/or its affiliates. | Audit log          | GPL            | ON          |
+-------------+----------------+---------------+-------------+---------------------+----------------+------------------------+------------------------------------+--------------------+----------------+-------------+

ubuntu@t-ubuntu1604-64:~$ sudo percona_server_admin --user=root --password=Tomi1234! --disable-audit
Checking Audit Log plugin status...
INFO: Audit Log plugin is installed.

Uninstalling Audit Log plugin...
INFO: Successfully uninstalled Audit Log plugin.
```

**PAM plugin:**
```
ubuntu@t-ubuntu1604-64:~$ sudo percona_server_admin --user=root --password=Tomi1234! --enable-pam
Checking PAM plugin status...
INFO: PAM Authentication plugin is not installed.

Installing PAM Authentication plugin...
INFO: Successfully installed PAM Authentication plugin.

ubuntu@t-ubuntu1604-64:~$ mysql --user=root --password=Tomi1234! -e "select * from information_schema.plugins where plugin_name like '%auth%';"
mysql: [Warning] Using a password on the command line interface can be insecure.
+-------------+----------------+---------------+----------------+---------------------+----------------+------------------------+---------------+---------------------------+----------------+-------------+
| PLUGIN_NAME | PLUGIN_VERSION | PLUGIN_STATUS | PLUGIN_TYPE    | PLUGIN_TYPE_VERSION | PLUGIN_LIBRARY | PLUGIN_LIBRARY_VERSION | PLUGIN_AUTHOR | PLUGIN_DESCRIPTION        | PLUGIN_LICENSE | LOAD_OPTION |
+-------------+----------------+---------------+----------------+---------------------+----------------+------------------------+---------------+---------------------------+----------------+-------------+
| auth_pam    | 0.1            | ACTIVE        | AUTHENTICATION | 1.1                 | auth_pam.so    | 1.7                    | Percona, Inc. | PAM authentication plugin | GPL            | ON          |
+-------------+----------------+---------------+----------------+---------------------+----------------+------------------------+---------------+---------------------------+----------------+-------------+
ubuntu@t-ubuntu1604-64:~$ sudo percona_server_admin --user=root --password=Tomi1234! --disable-pam
Checking PAM plugin status...
INFO: PAM Authentication plugin is installed.

Uninstalling PAM Authentication plugin...
INFO: Successfully uninstalled PAM Authentication plugin.
```

**MySQLX:**
```
ubuntu@t-ubuntu1604-64:~$ sudo percona_server_admin --user=root --password=Tomi1234! --enable-mysqlx
Checking MySQL X plugin status...
INFO: MySQL X plugin is not installed.

Installing MySQL X plugin...
INFO: Successfully installed MySQL X plugin.

ubuntu@t-ubuntu1604-64:~$ mysql --user=root --password=Tomi1234! -e "select * from information_schema.plugins where plugin_name like '%mysqlx%';"
mysql: [Warning] Using a password on the command line interface can be insecure.
+-------------+----------------+---------------+-------------+---------------------+----------------+------------------------+---------------+--------------------+----------------+-------------+
| PLUGIN_NAME | PLUGIN_VERSION | PLUGIN_STATUS | PLUGIN_TYPE | PLUGIN_TYPE_VERSION | PLUGIN_LIBRARY | PLUGIN_LIBRARY_VERSION | PLUGIN_AUTHOR | PLUGIN_DESCRIPTION | PLUGIN_LICENSE | LOAD_OPTION |
+-------------+----------------+---------------+-------------+---------------------+----------------+------------------------+---------------+--------------------+----------------+-------------+
| mysqlx      | 1.0            | ACTIVE        | DAEMON      | 50717.0             | mysqlx.so      | 1.7                    | Oracle Corp   | X Plugin for MySQL | GPL            | ON          |
+-------------+----------------+---------------+-------------+---------------------+----------------+------------------------+---------------+--------------------+----------------+-------------+

ubuntu@t-ubuntu1604-64:~$ sudo percona_server_admin --user=root --password=Tomi1234! --disable-mysqlx
Checking MySQL X plugin status...
INFO: MySQL X plugin is installed.

Uninstalling MySQL X plugin...
INFO: Successfully uninstalled MySQL X plugin.
```

**--config-file option**
```
ubuntu@t-ubuntu1604-64:~$ cat config
USER=root
PASSWORD=Tomi1234!
ENABLE_AUDIT=1
ENABLE_PAM=1
ENABLE_MYSQLX=1

ubuntu@t-ubuntu1604-64:~$ sudo percona_server_admin --config-file=config
Checking Audit Log plugin status...
INFO: Audit Log plugin is not installed.

Checking PAM plugin status...
INFO: PAM Authentication plugin is not installed.

Checking MySQL X plugin status...
INFO: MySQL X plugin is not installed.

Installing Audit Log plugin...
INFO: Successfully installed Audit Log plugin.

Installing PAM Authentication plugin...
INFO: Successfully installed PAM Authentication plugin.

Installing MySQL X plugin...
INFO: Successfully installed MySQL X plugin.
```

tplavcic commented:
Had a discussion with @dbpercona about how this will work with packaging. Since this script supports both  TokuDB and MyRocks and also some other plugins it will have to be packaged in the main server package but then it needs to make some checks for libraries if you for example want to install MyRocks/TokuDB but the libraries are not installed in the system.
So please wait with further review/merge until I push changes for these library checks.

tplavcic commented:
Ok, I made few changes:
1. the script no longer needs to be run as root unless you're installing TokuDB/TokuBackup (because of messing around with config file etc.)
2. if you get an error while trying to get a list of plugins you'll get output so you know what exact error it is
3. `${SCRIPT_PWD}/lib` is replaced with `${SCRIPT_PWD%/*}/lib` because the first one is not pointing correctly to libraries in binary tarball installation - eg. you would get `abs_basedir/bin/lib/..` while it should be `abs_basedir/lib/...` - I think this is a bug in the old script also
4. if ha_tokudb.so or ha_rocksdb.so are not found in any of the usual directories you'll get an error before even trying to install them so you won't need to investigate

Here's how it looks:
**RocksDB installation from tarball**
```
plavi@zoidberg  ~/sandboxes/binaries/5.7.17/bin  ./percona_server_admin -umsandbox --password=msandbox --socket=/tmp/mysql_sandbox5717.sock --enable-rocksdb
Checking if RocksDB plugin is available for installation ...
INFO: ha_rocksdb.so library for RocksDB found in /home/plavi/sandboxes/binaries/5.7.17/lib/mysql/plugin/ha_rocksdb.so.

Checking RocksDB engine plugin status...
INFO: RocksDB engine plugin is not installed.

Installing RocksDB engine...
INFO: Successfully installed RocksDB engine plugin.
```

**Root message**
```
vagrant@t-ubuntu14:~$ percona_server_admin -uroot --password=Tomi1234! --enable-tokudb
ERROR: For TokuDB install/uninstall this script must be run as root!
```

**TokuDB install from packages**
```
vagrant@t-ubuntu14:~$ sudo percona_server_admin -uroot --password=Tomi1234! --enable-tokudb                                                                                                                                            
Checking if TokuDB plugin is available for installation ...
INFO: ha_tokudb.so library for TokuDB found in /usr/lib/mysql/plugin/ha_tokudb.so.

Checking if Percona Server is running with jemalloc enabled...
INFO: Percona Server is running with jemalloc enabled.

Checking transparent huge pages status on the system...
INFO: Transparent huge pages are currently disabled on the system.

Checking if thp-setting=never option is already set in config file...
INFO: Option thp-setting=never is not set in the config file.
      (needed only if THP is not disabled permanently on the system)

Checking TokuDB engine plugin status...
INFO: TokuDB engine plugin is not installed.

Adding thp-setting=never option into /etc/mysql/my.cnf
INFO: Successfully added thp-setting=never option into /etc/mysql/my.cnf

Installing TokuDB engine...
INFO: Successfully installed TokuDB engine plugin.
```

**Error if ha_tokudb.so not available for installation**
```
vagrant@t-ubuntu14:~$ sudo percona_server_admin -uroot --password=Tomi1234! --enable-tokudb
Checking if TokuDB plugin is available for installation ...
ERROR: Cannot find ha_tokudb.so library for TokuDB installation.
Make sure you have TokuDB package installed or if running from binary tarball that this library is available.
```

For now I think this is it - IMHO it can be reviewed and merged - more testing will need to come later anyway.

george-lorch commented:
I _really_ like the design of this, it is simple and obvious and explains failures in way that a complete newb can easily google and understand the issue. 

We will need good documentation, I strongly suggest we get some docs done and reviewed by experts and EMT before we fully adopt the script.

tplavcic commented:
Opened DOC-401 for documentation changes.

vadimtk commented on scripts/CMakeLists.txt:
```diff
@@ -452,6 +452,10 @@ ELSE()
   ENDIF()
 
+  IF(WITH_ROCKSDB)
+    LIST(APPEND BIN_SCRIPTS percona_server_admin)
```
maybe we should use a shorter name? `ps_admin` ?

george-lorch replied:
In a previous discussion, I believe PZ preferred the the more verbose percona_server to avoid any ambiguity with other tools and also as a bit of a marketing thing to keep the full Percona brand recognizable. Not disagreeing with you, just putting the info out there.

tplavcic commented:
I have added following message into ps_tokudb_admin: `WARNING: This script is deprecated and will be removed in 8.0. You can use ps_admin script which has more functionality.`
I have also removed RocksDB options from help screen, but I have left the code for RocksDB in the script so it can be used in testing.

tplavcic commented:
@EvgeniyPatlan There are few minor changes in some packaging files. Please check if it's ok.

EvgeniyPatlan commented:
lgtm

tplavcic commented:
As per Vadim's request the tool is now renamed to `ps-admin`.

tplavcic commented:
I have added bats tests for ps-admin here: https://github.com/Percona-QA/package-testing/tree/master/bats
Once `ps-admin` is merged those tests will need to start running in our package testing automation.

Here's how it looks:
```
UNIT TESTS:
vagrant@ubuntu1610-64:/vagrant/bats$ bats ps-admin_unit.bats
 ✓ run ps-admin without any arguments
 ✓ display ps-admin help screen
 ✓ run ps-admin with wrong option
 ✓ run ps-admin --config-file with missing parameter
 ✓ run ps-admin --user with missing parameter
 ✓ run ps-admin --password with missing parameter
 ✓ run ps-admin --socket with missing parameter
 ✓ run ps-admin --host with missing parameter
 ✓ run ps-admin --port with missing parameter
 ✓ run ps-admin --defaults-file with missing parameter
 ✓ test message for installing TokuDB if user is not root

11 tests, 0 failures

INTEGRATION TESTS:
vagrant@ubuntu1610-64:/vagrant/bats$ sudo bats ps-admin_integration.bats
 ✓ install QRT plugin
 ✓ uninstall QRT plugin
 ✓ install Audit Log plugin
 ✓ uninstall Audit Log plugin
 ✓ install PAM plugin
 ✓ uninstall PAM plugin
 ✓ install MySQL X plugin
 ✓ uninstall MySQL X plugin
 ✓ install TokuDB plugin
 ✓ uninstall TokuDB plugin
 ✓ install TokuBackup plugin
 ✓ uninstall TokuDB and TokuBackup plugin
 ✓ install ALL plugins at once
 ✓ uninstall ALL plugins at once

14 tests, 0 failures
```
